File: ai-service/app/graph/state.py
from typing import TypedDict, List, Dict, Any, Optional

class TripState(TypedDict, total=False):
    # User Inputs & Derived Query
    user_id: Optional[str]
    user_request: str
    destination: str
    starting_city: str
    duration: int
    budget: float
    travelers: int
    interests: List[str]
    travel_style: str
    missing_fields: List[str]
    input_currency: str
    display_currency: str

    # Injected User Preferences Memory
    user_long_term_preferences: Dict[str, Any]

    # Agent Research Outputs
    places_found: List[Dict[str, Any]]
    weather_forecast: Dict[str, Any]
    transport_options: List[Dict[str, Any]]

    # Travel Intelligence Agent Analytics Output
    travel_intelligence: Dict[str, Any]

    # Budget Allocation & Currency Normalization
    budget_breakdown: Dict[str, Any]
    total_estimated_cost: float
    currency_rates: Dict[str, float]

    # Generated Day-by-Day Itinerary
    itinerary: Dict[str, Any]

    # Human-in-the-Loop Interruption State
    requires_human_input: bool
    human_prompt_options: List[Dict[str, Any]]
    clarification_prompt: str
    user_decision: Optional[str]

    # Cyclic Re-Planning & Validation State
    validation_passed: bool
    validation_issues: List[str]
    validation_feedback: str
    retry_count: int

    # Execution Trace Logs
    agent_logs: List[Dict[str, Any]]

File: server/src/graph/state.js
import { Annotation } from '@langchain/langgraph';

/**
 * TripStateAnnotation defines the centralized, stateful graph schema
 * passed between all specialized LangGraph agents in WanderWave.
 */
export const TripStateAnnotation = Annotation.Root({
  // User Input & Parsing State
  userId: Annotation({
    reducer: (x, y) => y ?? x,
    default: () => '',
  }),
  userRequest: Annotation({
    reducer: (x, y) => y ?? x,
    default: () => '',
  }),
  destination: Annotation({
    reducer: (x, y) => y ?? x,
    default: () => '',
  }),
  startingCity: Annotation({
    reducer: (x, y) => y ?? x,
    default: () => 'Delhi',
  }),
  duration: Annotation({
    reducer: (x, y) => y ?? x,
    default: () => 3,
  }),
  budget: Annotation({
    reducer: (x, y) => y ?? x,
    default: () => 20000,
  }),
  travelers: Annotation({
    reducer: (x, y) => y ?? x,
    default: () => 2,
  }),
  interests: Annotation({
    reducer: (x, y) => y ?? x,
    default: () => [],
  }),
  travelStyle: Annotation({
    reducer: (x, y) => y ?? x,
    default: () => 'Balanced',
  }),
  missingFields: Annotation({
    reducer: (x, y) => y ?? x,
    default: () => [],
  }),

  // Long-Term User Preferences Memory Injected
  userLongTermPreferences: Annotation({
    reducer: (x, y) => y ?? x,
    default: () => ({}),
  }),

  // Research Outputs (Populated by Day 5 Agents)
  placesFound: Annotation({
    reducer: (x, y) => y ?? x,
    default: () => [],
  }),
  weatherForecast: Annotation({
    reducer: (x, y) => y ?? x,
    default: () => ({}),
  }),
  transportOptions: Annotation({
    reducer: (x, y) => y ?? x,
    default: () => [],
  }),

  // Budget Allocation (Populated by Day 6 Agent)
  budgetBreakdown: Annotation({
    reducer: (x, y) => y ?? x,
    default: () => ({}),
  }),
  totalEstimatedCost: Annotation({
    reducer: (x, y) => y ?? x,
    default: () => 0,
  }),

  // Generated Plan & Loop Control (Populated by Day 7 & 8 Agents)
  itinerary: Annotation({
    reducer: (x, y) => y ?? x,
    default: () => ({}),
  }),
  validationErrors: Annotation({
    reducer: (x, y) => y ?? x,
    default: () => [],
  }),
  isValid: Annotation({
    reducer: (x, y) => y ?? x,
    default: () => false,
  }),
  retryCount: Annotation({
    reducer: (x, y) => y ?? x,
    default: () => 0,
  }),

  // Human-in-the-Loop Interruption State
  requiresHumanInput: Annotation({
    reducer: (x, y) => y ?? x,
    default: () => false,
  }),
  humanPromptOptions: Annotation({
    reducer: (x, y) => y ?? x,
    default: () => [],
  }),
  userDecision: Annotation({
    reducer: (x, y) => y ?? x,
    default: () => null,
  }),
  
  // Execution Log Trace for UI Real-Time Tracker
  agentLogs: Annotation({
    reducer: (x, y) => (x || []).concat(y || []),
    default: () => [],
  }),
});
